activerecord/lib/active_record/attribute_methods.rb

```diff
@@ -44,11 +44,67 @@ def initialize_generated_modules # :nodoc:
         @generated_attribute_methods = const_set(:GeneratedAttributeMethods, GeneratedAttributeMethods.new)
         private_constant :GeneratedAttributeMethods
         @attribute_methods_generated = false
+        @alias_attributes_mass_generated = false
         include @generated_attribute_methods
 
         super
       end
 
+      def alias_attribute(new_name, old_name)
+        super
+
+        if @alias_attributes_mass_generated
+          ActiveSupport::CodeGenerator.batch(generated_attribute_methods, __FILE__, __LINE__) do |code_generator|
+            generate_alias_attribute_methods(code_generator, new_name, old_name)
+          end
+        end
+      end
+
+      def eagerly_generate_alias_attribute_methods(_new_name, _old_name) # :nodoc:
+        # alias attributes in Active Record are lazily generated
+      end
+
+      def generate_alias_attributes # :nodoc:
+        return if @alias_attributes_mass_generated
+
+        generated_attribute_methods.synchronize do
+          return if @alias_attributes_mass_generated
+          ActiveSupport::CodeGenerator.batch(generated_attribute_methods, __FILE__, __LINE__) do |code_generator|
+            attribute_aliases.each do |new_name, old_name|
+              generate_alias_attribute_methods(code_generator, new_name, old_name)
+            end
+          end
+
+          @alias_attributes_mass_generated = true
+        end
+      end
+
+      def alias_attribute_method_definition(code_generator, pattern, new_name, old_name)
+        method_name = pattern.method_name(new_name).to_s
+        target_name = pattern.method_name(old_name).to_s
+        parameters = pattern.parameters
+        old_name = old_name.to_s
+
+        method_defined_within = method_defined_within?(target_name, self)
+        manually_defined_on_class = method_defined_within && self.instance_method(target_name).owner == self
+
+        if manually_defined_on_class
+          aliased_method_redefined_as_well = method_defined_within?(method_name, self)
+          return if aliased_method_redefined_as_well
```

Comment:
The redefinition might `super`. For this case I think we might need to (ugh) define the alias method with the deprecation warning inside it.

As it's temporary, we can probably just stick an optional `body_prefix` kwarg onto AMo's `alias_attribute_method_definition` I guess?

Reply:
>The redefinition might super

Not an existing one as we were eagerly defining aliased reader directly on class instead of a module which means there was no `super`. Let me know if I'm missing another use-case but for a model like:
```ruby
class Car < ActiveRecord::Base
  alias_attribute :make_name, :make

  def make_name
    super
  end
end
```
Calling `Car.new(make: "A").make_name` raises `NoMethodError: super: no superclass method `make_name' for #<Car`

And if someone addresses the deprecation warning by redefining the method with using `super` - its going to be on them if something breaks. We could improve the deprecation message though